### DISCUSSION TITLE
RAW hazard handling in olympia

### DISCUSSION THREAD

**Discussion by avinashmehtadelhi** [riscv-software-src/riscv-perf-model#72] [Q&A]
Hi
Wanted to know more about how olympia handles RAW hazards.
PFA log from dhrystone trace
Following are adjacent 2 instructions
uid: 90 1147a 'sub	x10,x14,x15'
uid: 91 1147e 'andi	x11,x10, +0xff'
(RAW hazard x10)
but both of them enter the ALU in clock 132, exit ALU in clock 133 and retire in 137
Is this correct behaviour?
[dhry_big_1k.log.verbose.txt](https://github.com/riscv-software-src/riscv-perf-model/files/12195335/dhry_big_1k.log.verbose.txt)

**Comment by klingaard** [comment]
How old is the version of olympia that you're running?  The latest master of olympia has instruction dependency tracking (add in PR https://github.com/riscv-software-src/riscv-perf-model/pull/51).  

You should see new logging messages like "Sending to issue queue..." and "Instruction NOT ready..." in the execution pipe.

**Comment by avinashmehtadelhi** [comment]
I update the repo, but the issue still seems to be there. PFA updated logs
[dhry_big_1k.log.verbose.log](https://github.com/riscv-software-src/riscv-perf-model/files/12216889/dhry_big_1k.log.verbose.log)
![raw hazard](https://github.com/riscv-software-src/riscv-perf-model/assets/16150599/65863d65-9b5f-4a1c-9e75-027d1b8aa816)

**Reply by aarongchan** [reply]
Not sure why you're seeing that RAW hazard being ignored by rename. When I run on master I see that UID 91 gets marked as "not ready" and doesn't get scheduled until UID 90 is retired.
I ran on the dhrystone trace for 1000 instructions (./olympia -i1k)

SS from [all_messages.log.verbose.txt](https://github.com/riscv-software-src/riscv-perf-model/files/12217292/all_messages.log.verbose.txt)

<img width="912" alt="image" src="https://github.com/riscv-software-src/riscv-perf-model/assets/22247127/ce8a4997-bf77-4052-8216-c4050be69f76">


<img width="932" alt="image" src="https://github.com/riscv-software-src/riscv-perf-model/assets/22247127/dbc8b2f3-91fa-4048-9c05-d024882dcced">

**Comment by ghost** [comment]
Hey @AaronGChan did you try big-core?  I'm seeing the same thing as @avinashmehtadelhi:
```
./olympia --arch big_core traces/dhry_riscv.zstf  -l top.*.*.execute info exe.out -i 100
14:31 $ grep -e "uid: 90" -e "uid: 91" exe.out 
{0000000137 00000137 top.cpu.core0.execute.alu2 info} getInstsFromDispatch_: Sending to issue queueuid: 90 DISPATCHED 1147a pid: 90 'sub	x10,x14,x15' 
{0000000137 00000137 top.cpu.core0.execute.alu3 info} getInstsFromDispatch_: Sending to issue queueuid: 91 DISPATCHED 1147e pid: 91 'andi	x11,x10, +0xff' 
{0000000137 00000137 top.cpu.core0.execute.alu2 info} issueInst_: Executing: uid: 90  SCHEDULED 1147a pid: 90 'sub	x10,x14,x15'  for 138
{0000000137 00000137 top.cpu.core0.execute.alu3 info} issueInst_: Executing: uid: 91  SCHEDULED 1147e pid: 91 'andi	x11,x10, +0xff'  for 138
{0000000138 00000138 top.cpu.core0.execute.alu2 info} completeInst_: Completing inst: uid: 90  COMPLETED 1147a pid: 90 'sub	x10,x14,x15' 
{0000000138 00000138 top.cpu.core0.execute.alu3 info} completeInst_: Completing inst: uid: 91  COMPLETED 1147e pid: 91 'andi	x11,x10, +0xff' 

```

**Reply by aarongchan** [reply]
Ah I see the issue now, I think I have a fix for it. We weren't clearing the scoreboard bits when allocating a PRF.

**Reply by avinashmehtadelhi** [reply]
Thanks for the fix
